File: 0020-valid-parentheses/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_brackets = "([{"
        close_brackets = ")]}"
        stack = []

        for c in s:

            # If character is an opening bracket
            if c in open_brackets:
                stack.append(c)
                continue

            # If character is a closing bracket
            if c in close_brackets:
                pos = close_brackets.index(c)  # corresponding opening index

                if not stack:        # nothing to match
                    return False

                if stack[-1] != open_brackets[pos]:  # mismatch
                    return False

                stack.pop()
                continue

            # invalid character (optional)
            return False

        return len(stack) == 0


if __name__ == "__main__":
    sol = Solution()
    print(sol.isValid("()"))          # True
    print(sol.isValid("()[]{}"))      # True
    print(sol.isValid("(]"))          # False
    print(sol.isValid("([)]"))        # False
    print(sol.isValid("{[]}"))        # True
    print(sol.isValid("]"))           # False
    print(sol.isValid("("))           # False

File: 0020-valid-parentheses/main.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        string open = "([{";
        string close = ")]}";
        stack<char> st;

        for (char c : s) {
            size_t posOpen = open.find(c);
            if (posOpen != string::npos) {
                st.push(c);
                continue;
            }

            size_t posClose = close.find(c);
            if (posClose != string::npos) {
                if (st.empty()) return false;               
                if (st.top() != open[posClose]) return false; 
                st.pop();
                continue;
            }

            return false;
        }

        return st.empty();
    }
};
